include/bm/bm_sim/actions.h

```diff
@@ -96,6 +97,42 @@ class NamedCalculation;
 class MeterArray;
 class CounterArray;
 
+// Internal representation of actions which are executed
+// periodically at a given interval
+struct PeriodicAction {
+  const char *name;
+  std::function<void()> fn;
+  std::chrono::milliseconds interval_ms;
+  unsigned int interval_ticks;
+};
+
+// Contains the list of the periodically executed actions
+class PeriodicActionsList {
+ public:
+  static PeriodicActionsList *get_instance();
+  bool register_periodic(const char *name,
+                         std::function<void()> fn,
+                         std::chrono::milliseconds interval);
+
+  // To be called continuously : waits for the next interval to
+  // elapse then calls the next registered function or set of functions.
+  // Returns `false` if no actions are registered
+  bool next();
+
+ private:
+  unsigned int tick;
+  std::vector<PeriodicAction> list_{};
```

Comment:
Wouldn't a priority queue be more appropriate for this? The actions can be ordered based on the next execution timestamp (last + period) and you can sleep until the next action (first element of the queue) is ready to be executed. You wouldn't need a min interval and you wouldn't need to iterate through the whole list of actions. I think you would get O(logN) complexity every time an action is executed.

Reply:
Absolutely true. Will update to use priority queue shortly.